### PR TITLE
Feat/add resource name to generated docs

### LINKED ISSUE #413: Add name column to resources and modules table
### What problem are you facing?

I can see the names of resources in the [terraform registry](https://registry.terraform.io/modules/tmknom/elasticache-redis/aws/latest?tab=resources) but cannot in terraform-docs.

```
aws_security_group.default
```

### How could terraform-docs help solve your problem?

I'd like a new resource name column added to the resources table or show the full name of each resource like in the terraform registry

### PR DESCRIPTION
**Pull request by edgarsandi**
#### TL;DR:
Fixes #413
Requires:  [pull request in plugin-sdk](https://github.com/terraform-docs/plugin-sdk/pull/20)

With this pull request, we add to the terraform-docs the ability to generate the name of resources with the resource name

Like the Resources tab in [terraform registry](https://registry.terraform.io/modules/tmknom/elasticache-redis/aws/latest?tab=resources) (as description in issue)

Following the approach already used for terraform providers, that repeat the providers that use alias

This pull request is as a draft because it needs the merge of the [pull request in plugin-sdk](https://github.com/terraform-docs/plugin-sdk/pull/20)

I have:

- [x] Read and followed terraform-docs' [contribution process].
- [x] All tests pass when I run `make test`.

##### `terraform-docs asciidoc document .`
```
== Resources

The following resources are used by this module:

- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity[aws_caller_identity.current]
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity[aws_caller_identity.ident]
- https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource[null_resource.foo]
- https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key[tls_private_key.baz]
```

##### `terraform-docs asciidoc table .`
```
== Resources

[cols="a",options="header,autowidth"]
|===
|Name
|https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity[aws_caller_identity.current]
|https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity[aws_caller_identity.ident]
|https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource[null_resource.foo]
|https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key[tls_private_key.baz]
|===
```

##### `terraform-docs json .`
```json
  "resources": [
    {
      "type": "caller_identity",
      "name": "current",
      "providerName": "aws",
      "provicerSource": "hashicorp/aws",
      "mode": "data",
      "version": "latest"
    },
    {
      "type": "caller_identity",
      "name": "ident",
      "providerName": "aws",
      "provicerSource": "hashicorp/aws",
      "mode": "data",
      "version": "latest"
    },
    {
      "type": "resource",
      "name": "foo",
      "providerName": "null",
      "provicerSource": "hashicorp/null",
      "mode": "managed",
      "version": "latest"
    },
    {
      "type": "private_key",
      "name": "baz",
      "providerName": "tls",
      "provicerSource": "hashicorp/tls",
      "mode": "managed",
      "version": "latest"
    }
  ]
```

##### `terraform-docs markdown document .`
```markdown
## Resources

The following resources are used by this module:

- [aws_caller_identity.current](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity)
- [aws_caller_identity.ident](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity)
- [null_resource.foo](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource)
- [tls_private_key.baz](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key)
```

##### `terraform-docs markdown table .`
```markdown
## Resources

| Name |
|------|
| [aws_caller_identity.current](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity) |
| [aws_caller_identity.ident](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity) |
| [null_resource.foo](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource) |
| [tls_private_key.baz](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key) |
```

##### `terraform-docs pretty .`
```pretty
resource.aws_caller_identity.current (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity)
resource.aws_caller_identity.ident (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity)
resource.null_resource.foo (https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource)
resource.tls_private_key.baz (https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key)
```

##### `terraform-docs toml .`
```toml
[[resources]]
  type = "caller_identity"
  name = "current"
  providerName = "aws"
  providerSource = "hashicorp/aws"
  mode = "data"
  version = "latest"

[[resources]]
  type = "caller_identity"
  name = "ident"
  providerName = "aws"
  providerSource = "hashicorp/aws"
  mode = "data"
  version = "latest"

[[resources]]
  type = "resource"
  name = "foo"
  providerName = "null"
  providerSource = "hashicorp/null"
  mode = "managed"
  version = "latest"

[[resources]]
  type = "private_key"
  name = "baz"
  providerName = "tls"
  providerSource = "hashicorp/tls"
  mode = "managed"
  version = "latest"
```

##### `terraform-docs xml .`
```xml
  <resources>
    <resource>
      <type>caller_identity</type>
      <name>current</name>
      <providerName>aws</providerName>
      <providerSource>hashicorp/aws</providerSource>
      <mode>data</mode>
      <version>latest</version>
    </resource>
    <resource>
      <type>caller_identity</type>
      <name>ident</name>
      <providerName>aws</providerName>
      <providerSource>hashicorp/aws</providerSource>
      <mode>data</mode>
      <version>latest</version>
    </resource>
    <resource>
      <type>resource</type>
      <name>foo</name>
      <providerName>null</providerName>
      <providerSource>hashicorp/null</providerSource>
      <mode>managed</mode>
      <version>latest</version>
    </resource>
    <resource>
      <type>private_key</type>
      <name>baz</name>
      <providerName>tls</providerName>
      <providerSource>hashicorp/tls</providerSource>
      <mode>managed</mode>
      <version>latest</version>
    </resource>
  </resources>
```

##### `terraform-docs yaml .`
```yaml
resources:
  - type: caller_identity
    name: current
    providerName: aws
    providerSource: hashicorp/aws
    mode: data
    version: latest
  - type: caller_identity
    name: ident
    providerName: aws
    providerSource: hashicorp/aws
    mode: data
    version: latest
  - type: resource
    name: foo
    providerName: "null"
    providerSource: hashicorp/null
    mode: managed
    version: latest
  - type: private_key
    name: baz
    providerName: tls
    providerSource: hashicorp/tls
    mode: managed
    version: latest
```

[contribution process]: https://git.io/JtEzg

### CODE REVIEW

File: internal/format/templates/markdown_table.tmpl [hunk 1]
```diff
@@ -53,9 +53,9 @@
         |------|
         {{- range .Module.Resources }}
         {{ if eq (len .URL) 0 }}
-            | {{ .FullType }}
+            | {{ .FullType }}.{{ .Name }}
```
Comment (khos2ow) [inline internal/format/templates/markdown_table.tmpl:56]: I just noticed there's `|` missing at the end of this line.

```suggestion
            | {{ .FullType }}.{{ .Name }} |
```

File: internal/terraform/resource.go [hunk 2]
```diff
@@ -70,5 +72,5 @@ type resourcesSortedByType []*Resource
 func (a resourcesSortedByType) Len() int      { return len(a) }
 func (a resourcesSortedByType) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 func (a resourcesSortedByType) Less(i, j int) bool {
-	return a[i].FullType() < a[j].FullType() || (a[i].FullType() == a[j].FullType() && a[i].FullType()+a[i].Mode <= a[j].FullType()+a[j].Mode)
+	return a[i].FullType() < a[j].FullType() || (a[i].FullType() == a[j].FullType() && a[i].Mode <= a[j].Mode && a[i].Name <= a[j].Name)
```
Comment (khos2ow) [inline internal/terraform/resource.go:75]: Please add items to `sampleResources` and corresponding tests to `resource_test.go`

**Review by khos2ow — COMMENTED** [review]
Thank you @edgarsandi, this looks great, I only have couple comments below.

File: internal/terraform/resource.go (hunk 2 above)
Comment (edgarsandi) [reply internal/terraform/resource.go:75]: @khos2ow para alcançar esse objetivo dos testes, eu precisei usar os valores diretos do objeto ao invés de usar a função `FullType`, está ok?

Comment (khos2ow) [reply internal/terraform/resource.go:75]: Yes, that looks good!

**Review by khos2ow — COMMENTED** [review]
This looks great, please go ahead and squash your commits. Also you don't need to include `(xxx):` at the beginning of the commit message.

**Review by khos2ow — APPROVED** [review]
Thank you @edgarsandi, LGTM!

**Comment by sergei-ivanov** [comment]
I wonder if this could be made optional through configuration file and/or command line switch?
My feeling is that it exposes too many unnecessary _implementation_ details in what was supposed to be module _API_ docs.
This will look extremely busy in some of our modules that use lots of resources of the same type (e.g. `aws_security_group_rule`, `aws_network_acl_rule`, `aws_api_gateway_resource`, etc.)

**Comment by khos2ow** [comment]
That's a very good point @sergei-ivanov. Would you mind filing an issue instead of here to be able to reach broader audience for feedback?
